Added to the args. (ascii_chars, char_size)

gabe-burnette opened the pull request:
- Added a (-f) --full_characters bool which sets ascii_chars to full spectrum of characters ( " .:-=+*%#@1234567890ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz")
- Added (-r) --ascii_chars Use what characters you want to use in the image. (default: " .:-=+*%#@")
- Added (-a) --char_size Set the size of the characters. (default: 8)
- Added (-t) --threshold_disabled To disable the threshold.
- Also added short codes to sigma1 (-l) sigma2 (-m) to make them usable from the CLI. #9 
- Added these to prepare zig program for wasm.
- Updated readme.md and added args to test.

seatedro commented:
The characters string needs to be sorted in ascending order of their perceived brightness. `#` and `@` were chosen because they look brighter.

The alphabets are not that bright, so we would have to sort this. The other stuff is fine i think.

gabe-burnette commented:
Added a sorting function which takes in the ascii_chars string and sorts it based on the size of the individual char in font_bitmap. If a character doesn't exist in the font_bitmap, it is removed from the string.
Also added - `-d, --sorted_ovr`: Prevents sorting of the ascii_chars by size.

gabe-burnette commented:
Also added a random/stringSortTesting.zig which allows you to test inputs/output with the sorting.

seatedro commented on random/stringSortTesting.zig:
Don't create another entrypoint. We can just add new tests to a `tests/` folder in the root directory.

You can refer to the test i've used in the main.zig file.

gabe-burnette replied:
Ended up removing the test I added. Wasn't necessary.

seatedro commented on readme.md:
We'll hold off on the shorthand changes. This could be a bug in the dependency project https://github.com/Hejsil/zig-clap. Can you test it with zig's master branch on windows?

gabe-burnette replied:
It appears to be a bug. Added a disclaimer in the readme.md
`4. Using the long arguments on windows may or may not work. Please use the short arguments for now.`

seatedro commented on src/main.zig:
```diff
@@ -219,9 +229,77 @@ fn parseArgs(allocator: std.mem.Allocator) !Args {
+            }
+        } else blk: {
+            if (res.args.full_characters != 0) {
+                break :blk " .-:=+iltIcsv1x%7aejorzfnuCJT3*69LYpqy25SbdgFGOVXkPhmw48AQDEHKUZR@B#NW0M";
+            } else {
+                break :blk " .:-=+*%@#";
+            }
```
Any reason why we don't want globals?

gabe-burnette replied:
Added them as globals.

seatedro reviewed (commented):
Neat sort function! Need to test it out and see how images look in practice first though before merging

gabe-burnette commented:
Removed the one test. Added a disclaimer for the zig clap bug on windows to the readme. Added the one thing as globals.
Also, I had to flip the font_bitmap 180 degrees horizontally. The font is flipped. I didn't want to add a piece of code which flipped it in the program, which could slow it, so I just flipped in the const. Here's reference photos if you look closely.

Before flipping:
command: asciigen -i dingboard.png -s 0.25 -c -f
![before_flip](https://github.com/user-attachments/assets/7ee740a8-86b5-4e5f-8380-8eb006968466)

After flipping:
command: asciigen -i dingboard.png -s 0.25 -c -f
![after_flip](https://github.com/user-attachments/assets/fe908c65-9a6d-4785-96b1-a52b2177a229)

seatedro commented:
I'm down to merge everything except the short form args stuff. It doesn't feel right attaching random alphabets to words. Like, 
- `--scale` is `-s` that makes sense
- `--sigma1` being `-l` does not feel good.

I feel most of these values at default settings would still yield a decent looking image for the majority.

Remove the short form changes for now, keep the newly added long form args.

Can you try to find out why the short form arg is not working with zig-clap?

gabe-burnette commented:
Yeah, that's fine. They're random because I was just using them to test, because of the windows error lmao. Thought maybe a few of them would stick, but removed them. The only one to actually consider is -f, for full_characters. But, I removed it.
Updated the args and readme.
I'll keep looking to the short form arg to see if there's a solution or work around.

seatedro commented on src/main.zig:
```diff
@@ -181,6 +184,11 @@ fn parseArgs(allocator: std.mem.Allocator) !Args {
         \\-b, --brightness_boost <f32>   Brightness boost (default: 1.0)
+        \\    --full_characters          Uses full spectrum of characters in image.
+        \\    --ascii_chars <str>        Use what characters you want to use in the image. (default: " .:-=+*%#@")
+        \\    --sorted_ovr               Prevents sorting of the ascii_chars by size.
```
Can you rename this to --disable_sort or something?

gabe-burnette replied:
Renamed.

seatedro commented on src/main.zig:
```diff
@@ -7,10 +7,8 @@ const stb = @cImport({
-const CHAR_SIZE = 8;
-const THRESHOLD = 50;
+const default_characters = " .:-=+*%@#";
+const full_characters = " .-:=+iltIcsv1x%7aejorzfnuCJT3*69LYpqy25SbdgFGOVXkPhmw48AQDEHKUZR@B#NW0M";
```
I think the full characters don't look that good. The dots above the j overlap other pixels. Our default ascii characters should ideally be perfect.

Let's hold off on the full character list.

We should instead explore a Unicode character list instead.

seatedro replied:
![CleanShot 2024-09-02 at 23 56 13@2x](https://github.com/user-attachments/assets/633a2d56-7c4b-4dc4-994f-ee3552de8e45)

seatedro commented on readme.md:
```diff
@@ -34,9 +34,14 @@ run the program with the following options (the default zig install directory is
+   - `    --full_characters`: Uses full spectrum of characters in image.
+   - `    --ascii_chars <string>`: Use what characters you want to use in the image. (default: " .:-=+*%@#")
+   - `    --sorted_ovr`: Prevents sorting of the ascii_chars by size.
+   - `    --char_size <u8>`: Set the size of the characters. (default: 8)
```
This doesn't actually set the size of the characters. But the size of each block of image considered. It's kind of pointless changing this. 

See this image below: This is what happens if you set it to a value that isn't divisible by 8. 

![CleanShot 2024-09-02 at 23 49 08@2x](https://github.com/user-attachments/assets/5a6df04b-8b67-4796-91a0-bd53880fcc4a)

And when the value is set to something high like 24, the image loses becomes blocky:
![image](https://github.com/user-attachments/assets/a47d0fa8-7bda-4f51-bc97-a88a633c835e)

gabe-burnette commented:
Renamed sorted_ovr to disable_sort and char_size to block_size. 

I mainly added the full characters, block size, and threshold disabled because I was getting really cool results fucking around with the values, even if it is really glitchy, still makes cool results for some stuff.

I agree with the unicode. I think we should keep the full_characters for now, until we get a defined set. It doesn't have a short code, so it's not that important.

I added all of the 5 options I added as `advanced options (your image will break, probably, but you might get some cool results):` in the readme, with a disclaimer. Let me know if you want the wording changed.

seatedro commented:
that's fine, you're right. it does look cool.

merging
